js/src/builtin/streams/ReadableByteStreamController.cpp

```diff
@@ -4,232 +4,100 @@
-                                      JSMSG_READABLEBYTESTREAMCONTROLLER_BAD_CHUNKSIZE);
-            return nullptr;
-        }
-    }
-
-    // Step 12: Set this.[[autoAllocateChunkSize]] to autoAllocateChunkSize.
-    controller->setAutoAllocateChunkSize(autoAllocateChunkSize);
-
-    // Step 13: Set this.[[pendingPullIntos]] to a new empty List.
-    if (!StoreNewListInFixedSlot(cx, controller,
-                                 ReadableByteStreamController::Slot_PendingPullIntos)) {
-        return nullptr;
-    }
-
-    // Step 14: Let controller be this (implicit).
-
-    // Step 15: Let startResult be
-    //          ? InvokeOrNoop(underlyingSource, "start", « this »).
-    RootedValue startResult(cx);
-    RootedValue controllerVal(cx, ObjectValue(*controller));
-    if (!InvokeOrNoop(cx, underlyingByteSource, cx->names().start, controllerVal, &startResult)) {
-        return nullptr;
-    }
-
-    // Step 16: Let startPromise be a promise resolved with startResult:
-    RootedObject startPromise(cx, PromiseObject::unforgeableResolve(cx, startResult));
-    if (!startPromise) {
-        return nullptr;
-    }
-
-    RootedObject onStartFulfilled(cx, NewHandler(cx, ReadableStreamControllerStartHandler, controller));
-    if (!onStartFulfilled) {
-        return nullptr;
-    }
-
-    RootedObject onStartRejected(cx, NewHandler(cx, ControllerStartFailedHandler, controller));
-    if (!onStartRejected) {
-        return nullptr;
-    }
-
-    if (!JS::AddPromiseReactions(cx, startPromise, onStartFulfilled, onStartRejected)) {
-        return nullptr;
-    }
-
-    return controller;
-}
-
-#endif  // user-defined byte streams
-
-/**
- * Streams spec, 3.11.3.
- * new ReadableByteStreamController ( stream, underlyingByteSource,
- *                                    highWaterMark )
+ * Streams spec, 3.9.3.
+ * new ReadableByteStreamController( stream, underlyingSource, size,
+ *                                      highWaterMark )
  */
 bool ReadableByteStreamController::constructor(JSContext* cx, unsigned argc,
                                                Value* vp) {
-  // Step 1: Throw a TypeError exception.
+  // Step 1: Throw a TypeError.
   JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
                             JSMSG_BOGUS_CONSTRUCTOR,
                             "ReadableByteStreamController");
   return false;
 }
 
+const JSClass ByteStreamChunk::class_ = {"ByteStreamChunk",
+                                         JSCLASS_HAS_RESERVED_SLOTS(SlotCount)};
+
+ByteStreamChunk* ByteStreamChunk::create(JSContext* cx, HandleObject buffer,
+                                         uint32_t byteOffset,
+                                         uint32_t byteLength) {
+  Rooted<ByteStreamChunk*> chunk(cx,
+                                 NewBuiltinClassInstance<ByteStreamChunk>(cx));
+  if (!chunk) {
+    return nullptr;
+  }
+
+  chunk->setFixedSlot(Slot_Buffer, ObjectValue(*buffer));
```

Comment:
As we are dereferencing a HandleObject, I believe it could be a nullptr -- perhaps we could add a `MOZ_ASSERT(buffer)` before this call to ensure that we pass a valid pointer, or should we `return  nullptr;` in this situation?

```suggestion
  MOZ_ASSERT(buffer);
  chunk->setFixedSlot(Slot_Buffer, ObjectValue(*buffer));
```

Reply:
Extra asserts seem good to me!